LibraBFT/Impl/Properties/VotesOnceDirect.agda

```diff
@@ -104,65 +104,66 @@ module LibraBFT.Impl.Properties.VotesOnceDirect where
- 
-  peerCanSignEp≡ : ∀ {pid v v' pk s'}
-                   → PeerCanSignForPK s' v pid pk
-                   → v ^∙ vEpoch ≡ v' ^∙ vEpoch
-                   → PeerCanSignForPK s' v' pid pk
-  peerCanSignEp≡ (mkPCS4PK 𝓔₁ 𝓔id≡₁ 𝓔inSys₁ mbr₁ nid≡₁ pk≡₁) refl = (mkPCS4PK 𝓔₁ 𝓔id≡₁ 𝓔inSys₁ mbr₁ nid≡₁ pk≡₁)
-
-
-  -- This property does not hold!  The problem is that there is nothing constraining peerStates st
-  -- pid': pid' might be uninitialised, in which case PeerCanSignForPK (peerStates st pid') v' pid'
-  -- pk does not tell us anything. We cannot add a hypothesis that initialised st pid' ≡ initd,
-  -- because that's what we're trying to prove in one of the places this is used (note that pid in
-  -- msg∈pool⇒initd corresponds to pid' here).
-
-  -- I think we are running into exactly the thing that unwind helps with: it takes us back to the
-  -- the transition that first sends a signature, where we know it hasn't been sent before, and must be
-  -- sent my a step-msg, which only occurs if the sender is initialised, which we then carry forward
-  -- (see carrInit in Yasm.Properties).  That's not to say we can't succeed without unwind, of course,
-  -- but this challenge highlights the difference well.
+                    → MsgWithSig∈ pk (ver-signature sig) (msgPool pre)
+                    → PeerCanSignForPK pre v pid pk
+  peerCanSign-Msb4 r step (mkPCS4PK 𝓔₁ 𝓔id≡₁ (inGenInfo refl) mbr₁ nid≡₁ pk≡₁) pkH sig msv
+    = mkPCS4PK 𝓔₁ 𝓔id≡₁ (inGenInfo refl) mbr₁ nid≡₁ pk≡₁
 
-  peerCanSignPK-Inj :  ∀ {pid pid' s' outs pk v v'}{st : SystemState}
+  peerCanSignPK-Inj :  ∀ {pid pid' pk v v'}{st : SystemState}
                     → ReachableSystemState st
-                    → (stP : StepPeerState pid (msgPool st) (initialised st) (peerStates st pid) (s' , outs))
                     → Meta-Honest-PK pk
                     → PeerCanSignForPK st v' pid' pk
-                    → PeerCanSignForPK (StepPeer-post (step-honest stP)) v pid pk
+                    → PeerCanSignForPK st v pid pk
                     → v ^∙ vEpoch ≡ v' ^∙ vEpoch
                     → pid ≡ pid'
-  peerCanSignPK-Inj {pid} {pid'} {v = v} r stP pkH pcsv'Pre pcsvPost refl
-    with pid ≟ pid'
-  ...| yes refl = refl
-  ...| no pids≢
-     with step-peer (step-honest stP)
-  ...| theStep
-     with PeerCanSignForPK-stable r theStep pcsv'Pre
-  ...| pcsv'Post
-     with availEpochsConsistent (step-s r theStep) pcsv'Post pcsvPost
-  ...| refl = ⊥-elim (pids≢ (NodeId-PK-OK-injective (𝓔 pcsvPost)
-                                                    (PCS4PK⇒NodeId-PK-OK pcsvPost)
-                                                    ( PCS4PK⇒NodeId-PK-OK pcsv'Post)))
+  peerCanSignPK-Inj {pid} {pid'} r pkH pcs' pcs eid≡
+     with availEpochsConsistent r pcs' pcs
+  ...| refl
+     with NodeId-PK-OK-injective (𝓔 pcs) (PCS4PK⇒NodeId-PK-OK pcs) (PCS4PK⇒NodeId-PK-OK pcs')
+  ...| pids≡ = pids≡
+
 
   msg∈pool⇒initd : ∀ {pid pk v}{st : SystemState}
```

Comment:
This property won't hold as-is once we complete the initialisation work, because the signature in question may be derivable from the genesis info, in which case a cheat can send a message containing the signature, which it got from the genesis info.  I am wondering if we should include in the definition of `MsgWithSig∈` that the signature is not derivable from the genesis info, so we can eliminate that case any time we're dealing with a ``MsgWithSig∈`.  (I do realise that you were just following the advice I had in the comment you removed below, but this was before I had made progress on initialisation and genesis information.)

Reply:
Ok. Once you finish the initialisation work I will fix it.

Reply:
Ok thanks